include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -485,11 +489,132 @@ struct __radix_sort_scan_submitter<_RadixLocalScanName, __internal::__optional_k
+    using _TempStorageT =
+        sycl::accessor<::std::uint32_t, 1, sycl::access::mode::read_write, sycl::access::target::local>;
+
+    sycl::sub_group __sgroup;
+    ::std::uint32_t __self_lidx;
+    ::std::uint32_t __item_mask;
+    _AtomicT __atomic_peer_mask;
+
+    __peer_prefix_helper(sycl::nd_item<1> __self_item, _TempStorageT __lacc)
+        : __sgroup(__self_item.get_sub_group()), __self_lidx(__self_item.get_local_linear_id()),
+          __item_mask(~(~0u << (__self_lidx))), __atomic_peer_mask(__lacc[0])
+    {
+    }
+
+    ::std::uint32_t
+    __peer_contribution(_OffsetT& __new_offset_idx, _OffsetT __offset_prefix, bool __is_current_bucket)
+    {
+        // reset mask for each radix state
+        if (__self_lidx == 0)
+            __atomic_peer_mask.store(0U);
+        sycl::group_barrier(__sgroup);
+        // set local id's bit to 1 if the bucket value matches the radix state
+        __atomic_peer_mask.fetch_or(__is_current_bucket << __self_lidx);
+        sycl::group_barrier(__sgroup);
+        ::std::uint32_t __peer_mask_bits = __atomic_peer_mask.load();
+        ::std::uint32_t __sg_total_offset = sycl::popcount(__peer_mask_bits);
+
+        // get the local offset index from the bits set in the peer mask with index less than the work
+        // items's ID
+        __peer_mask_bits &= __item_mask;
+        __new_offset_idx |= __is_current_bucket * (__offset_prefix + sycl::popcount(__peer_mask_bits));
+        return __sg_total_offset;
+    }
+};
+
+template <typename _OffsetT>
+struct __peer_prefix_helper<_OffsetT, __peer_prefix_algo::scan_then_broadcast>
+{
+    using _TempStorageT = __empty_peer_temp_storage;
+
+    sycl::sub_group __sgroup;
+    ::std::uint32_t __sg_size;
+
+    __peer_prefix_helper(sycl::nd_item<1> __self_item, _TempStorageT)
+        : __sgroup(__self_item.get_sub_group()), __sg_size(__sgroup.get_local_linear_range())
+    {
+    }
+
+    ::std::uint32_t
+    __peer_contribution(_OffsetT& __new_offset_idx, _OffsetT __offset_prefix, bool __is_current_bucket)
+    {
+        ::std::uint32_t __sg_item_offset = __dpl_sycl::__exclusive_scan_over_group(
+            __sgroup, static_cast<::std::uint32_t>(__is_current_bucket), __dpl_sycl::__plus<::std::uint32_t>());
+
+        __new_offset_idx |= __is_current_bucket * (__offset_prefix + __sg_item_offset);
+        // the last scanned value may not contain number of all copies, thus adding __is_current_bucket
+        ::std::uint32_t __sg_total_offset =
+            __dpl_sycl::__group_broadcast(__sgroup, __sg_item_offset + __is_current_bucket, __sg_size - 1);
+
+        return __sg_total_offset;
+    }
+};
+
+template <typename _OffsetT>
+struct __peer_prefix_helper<_OffsetT, __peer_prefix_algo::subgroup_ballot>
+{
+    using _TempStorageT = __empty_peer_temp_storage;
+
+    sycl::sub_group __sgroup;
+    ::std::uint32_t __self_lidx;
+    sycl::ext::oneapi::sub_group_mask __item_sg_mask;
```

Comment:
Regarding the extension like this `sycl::ext::oneapi`. How it will work in case of "pure" SYCL?

Reply:
That's the goal of the atomic_fetch_or version above.